### ISSUE TITLE
PowerShell: First publish lost and for subsequent publishes missing end brackets '}' when using PowerShell to receive output

### ISSUE THREAD

**Issue by PramodKumarYadav** [jandelgado/rabtap#21]
**Summary:** 
I am using PowerShell to create jobs to set up the taps. 
Then I publish messages and receive job output. 

So, If I have three files to iterate from a folder to publish messages; What I see is that:
1. First files messages are "Lost"
2. Second files messages are "Received" but with a missing "}" for the last JSON msg.
3. Third (and subsequent) file, starts with "}{" and the last line misses "}" 

Note that this is not an issue when I run another PowerShell instance to Set up a tap and receive the output (i.e. only happens when I use the job mechanism). Due to this, I am not sure if this is entirely a PowerShell issue or something that you can fix in rabtap? In any case, I bring this to your notice so that you can try it at your end and atleast are aware of this issue. If something can be done to fix this, that would ofcourse be awesome! 

**Steps:**
The code that I am using is as below:

# Step01: Set up tap using PowerShell jobs
$pingblockForRawMessagesTap = {
            rabtap tap "Name.A.B.MyGateway\:GatewayMessage:" --json
        }
        
        Start-Job $pingblockForRawMessagesTap -Name tapGatewayMsg

# Step02: Publish messages by reading JSON files. 
Get-Content $inputMessages | rabtap pub  Name.A.B.MyGateway:GatewayMessage --json

# Step03: Receive output
Receive-Job -Name tapGatewayMsg | Out-File $outputRawMessages 

#Output of the first file will be lost.  Second have a missing '}' in end. and the next file starts as '}{' with each file missing '{' in the last line.

**Comment by PramodKumarYadav** [comment]
To correct: The issue happens even when I have another PowerShell instance open with a Tap (and not just for jobs). What I see is the issue happens when I redirect the output to file. When output is displayed on Powershell console (so not redirected), the output starts and ends properly. But when you redirect the output to file, that's where all the funny things start happening.

**Comment by PramodKumarYadav** [comment]
To give some more inputs, If I try to add the missing '}' from the captured messages from exchange and republish them, it results in an error. On the other hand, running the captured messages as-is (so without ending '}') works OK. 

So it appears that there is some correction being done in rabtap for this error (which if corrected outside, creates an issue)? I was trying to republish few json messages from my previously published messages when I encountered and figured out this issue. 

Hope this helps!

**Comment by jandelgado** [comment]
Hi, when using the `--json` option, rabtap produces (tap, sub) or consumes (pub) a stream of JSON messages. That's why you see e.g. `}{` in the output. That's just the ending of one and the beginning bracket of the following JSON message:
```
{
...message 1...
}{
...message 2...
} etc
```
So, this behaves as expected. As for your powershelgl question, I have no idea, I never used it.

**Comment by PramodKumarYadav** [comment]
Hello jandelgado, 

Thank you for your reply. Does that mean, I can close the issue or would you like to give it some time to investigate or discuss with someone who is more familiar with PowerShell? 

The impact this issue has on automation is that when I tap and redirect my output for further record/play/compare, the redirected output misses out on the last message having any closing bracket. And each file needs some work to be done before it can be used. 

So, If I publish three messages, and redirect each of them to a file, the last message never receives a closing bracket. the second would need work to be done both at start and end before it could be used and the first message is lost. May be other users in future may have similar issues while using rabtap as an automation tool for test automation? Nonetheless, if you recommend closing this issue, I will do so.

Thank you for support and creating an awesome tool. I love using it :).

**Comment by PramodKumarYadav** [comment]
To give some more inputs, When I run below command on windows cmd prompt,
rabtap tap "myExchange" --json > "C:\test\test08.txt"
It gives all properly formatted JSON output messages.

Issue happens when I run the exact same command (or any of its variant such as out-file, set-content etc) on Powershell. Then the weirdness starts. So maybe it has everything to do with PowerShell and nothing to do with rabtap itself? 

Do share your thoughts and then we can do as you suggest.

**Comment by jandelgado** [comment]
Given that your test on the windows cmd prompt works as expected, this is likely a PowerShell related problem and not a problem in rabtap. For another test, you could try to manually modify a pre-recorded JSON output, so that `}{` gets split onto two lines and check if this works. If this fixes the problem, I could change JSON output in rabtap. Otherwise I don't see how this can be fixed in rabtap and therefore will close the issue.

**Comment by PramodKumarYadav** [comment]
I tried the test but it did not work. So;
Steps:
->  I modified the pre-recorded output to put }{ on two lines. 
-> Put tap on exchange, redirecting output to a file.
-> Republished the modified message.
-> Checked output in the file.
Result: Closing bracket was still missing. 

Nonetheless, what I see is that when I do not redirect output to a file, ending bracket is visible on the prompt. Anyways I have raised a bug in PowerShell. Hopefully, they can recreate this issue at their end and do something about it. 

We can close it here. Thank you for your help and prompt responses!
